### PR TITLE
Canvas pages content picker

### PR DESCRIPTION
**Pull request by marcospri**
FE parts needed for the file picker.

Other than have  this written by someone who actually knows JS there a few small things:

- Ordering of the buttons on the file picker. Is it fine?
- Make the difference between Canvas Pages and Canvas Files more evident?
- A few of the text mention "files" when it should say "pages". The modal title, the text when no pages are found.

### CODE REVIEW

**Comment by acelaya** [comment]
I have force-pushed to this branch with the next changes.

- I have added a new `LMSPagePicker`, which is a simplified version of the file picker, without support for recursive loading (it loads a flat list and that's it). Once we have a better design, we can change it there.
- I have defined a new `Page` API type, and a common `Resource` type which both `File` and `Page` extend from.
- I have refactored the `FileList` component, and it is now a `ResourceList`, so that both `LMSFilePicker` and `LMSPagePicker` can use it. It is generic and can handle any object that implements `Resource`.
- I have added a new check to `ResourceList` so that it uses a different icon when the resource has `type: 'Page'`. Right now the BE returns pages with `type: 'File'`, so this is not immediately ovbious.

I haven't added tests for `LMSPagePicker` yet. Will wait for @robertknight to give this a first pass.

File: lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+ * Renders a helpful message with a link to documentation when there are no
+ * pages.
+ *
+ * @todo set useful link
+ */
+function NoPagesMessage() {
+  return (
+    <div>
+      There are no pages in this instance.{' '}
+      <a href={'TODO'} target="_blank" rel="noreferrer">
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx:27]: We need to add a proper link here like we do with files.

File: lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+      } catch (error) {
+        if (isAuthorizationError(error)) {
+          setDialogState({ state: 'authorizing', isRetry: isReload });
+        } else {
+          setDialogState({ state: 'error', error });
+        }
+      }
+      setInitialFetch(false);
+    },
+    [loadPagesToDisplay]
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx:153]: These two function could probably be merged. In the file picker it makes more sense to have them separated as the logic is more complex.

File: lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx (hunk 1 above)
Comment (marcospri) [reply lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx:27]: I'll create a board for this soon and I'll a draft about this.

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx [hunk 3]
```diff
@@ -164,6 +169,11 @@ export default function ContentSelector({
     onSelectContent({ type: 'file', file });
   };
 
+  const selectCanvasPage = (page: Page) => {
+    cancelDialog();
+    onSelectContent({ type: 'url', url: page.id });
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/ContentSelector.tsx:174]: Selecting the page itself doesn't work yet. That's out of this PR scope.

We probably need to change this later, once aligned with the BE.

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx [hunk 4]
```diff
@@ -333,9 +354,20 @@ export default function ContentSelector({
             onClick={() => selectDialog('canvasFile')}
             title="Select PDF from Canvas"
           >
-            Canvas
+            Canvas File
```
[the remark this answers is not in the document]
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/ContentSelector.tsx:357]: I think Canvas File makes sense.

**Comment by robertknight** [comment]
The actual ordering here is alphabetical left-to-right, top-to-bottom. My natural inclination was to read it as columns of a table though, in which case the buttons look jumbled up.

File picker ordering

**Comment by robertknight** [comment]
After choosing a page, we should ideally display something like "Canvas page: {title}" as the assignment content, rather than the URL we generate for internal use.

Canvas page display

File: lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx (hunk 1 above)
Comment (marcospri) [reply lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx:27]: Created here: https://github.com/orgs/hypothesis/projects/126?pane=issue&itemId=40446950

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx (hunk 3 above)
Comment (marcospri) [reply lms/static/scripts/frontend_apps/components/ContentSelector.tsx:174]: This should work now for the most basic cases in `main`.

**Comment by acelaya** [comment]
> After choosing a page, we should ideally display something like "Canvas page: {title}" as the assignment content, rather than the URL we generate for internal use.
> 
> Canvas page display

Changed

File: lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx [hunk 5]
```diff
@@ -0,0 +1,300 @@
+import {
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/LMSPagePicker.tsx:1]: Aside from the difference in which API calls are made and the terminology presented to the user, are there any other differences between files and pages? If not, this component and its tests are pretty complex and I'd probably lean towards using a single implementation until we have a clear reason to treat them differently.

**Review by robertknight — COMMENTED** [review]
I gave this a test in Canvas. The main feedback I have looking through the code is that the logic for dealing with authorization in `LMSFilePicker` and `LMSPagePicker` is quite involved and I wonder if it might be preferable to have only a single implementation and set of tests if possible. Other than the UI text, and that subfolders are currently not used for pages, are there any other differences between these types that require a separate component to render them?

**Comment by acelaya** [comment]
> I gave this a test in Canvas. The main feedback I have looking through the code is that the logic for dealing with authorization in `LMSFilePicker` and `LMSPagePicker` is quite involved and I wonder if it might be preferable to have only a single implementation and set of tests if possible. Other than the UI text, and that subfolders are currently not used for pages, are there any other differences between these types that require a separate component to render them?

Thanks Rob!

I think the main user-facing differences are the dialog title and the icons, but the LMSFilePicker would work (Marcos already proved this).

The difference in icon should be addressed by the FileList -> ResourceList refactoring, which involves a bunch of name changes, and the new icon when `type` is `Page`, but appart from that it's the same as before.

As per the title, it could be addressed via a simple prop.

So long story short, I'll try to simplify this as much as possible and use LMSFilePicker for now. Will let you guys know if I see anything else that makes this approach not viable.

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx [hunk 6]
```diff
@@ -285,10 +291,10 @@ export default function LMSFilePicker({
     if (!file) {
       return;
     }
-    if (!file.type || file.type === 'File') {
-      onSelectFile(file);
-    } else {
+    if (file.type === 'Folder') {
       onChangePath(file);
+    } else {
+      onSelectFile(file);
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx:297]: I flipped this condition so that resources with type `Page` work transparently for now, without having to explicitly reference pages here for now.

**Comment by acelaya** [comment]
@robertknight I have simplified this PR as suggested. These are the changes I have kept:

* I have defined a new `Page` API type, and a common `Resource` type which both `File` and `Page` extend from.
* I have refactored the `FileList` component, and it is now a `ResourceList` that can handle both `File` and `Page` objects. The only effective difference with the `FileList` (other than renaming references from "file" to "resource") is that it renders a different icon for `Page` entries.

**Comment by robertknight** [comment]
This authorization message will either need to be customized according to the resource type:

Auth message

**Comment by robertknight** [comment]
Pages are currently not sorted by name, which will make find an entry difficult if there are a large number. Is this handled on the frontend or backend for files?

Page sort order

File: lms/static/scripts/frontend_apps/api-types.ts [hunk 7]
```diff
@@ -30,6 +34,10 @@ export type File = {
   children?: File[];
 };
 
+export type Page = Resource & {
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/api-types.ts:37]: This could use a comment to indicate what a "Page" is. This could be something along the lines of "Canvas Pages or similar resources."

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx [hunk 8]
```diff
@@ -164,6 +168,15 @@ export default function ContentSelector({
     onSelectContent({ type: 'file', file });
   };
 
+  const selectCanvasPage = (page: File) => {
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/ContentSelector.tsx:171]: This argument has the type `File` although the resource is actually a page. This could be handled by:

1. Adding a separate `onSelectPage` callback to `LMSFilePicker`
2. Making the `onSelectFile` argument `File | Page` and then using a cast to convert it to the expected type
3. Making `LMSFilePicker` generic over the type of resource and then `onSelectFile` would emit the corresponding kind of resource.

Of these (2) is probably the easiest.

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx [hunk 9]
```diff
@@ -75,6 +75,11 @@ type LMSFilePickerProps = {
    * Render path breadcrumbs and allow sub-folder navigation?
    */
   withBreadcrumbs?: boolean;
+
+  /**
+   * An alternative title for the modal dialog. Defaults to "Select file"
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx:80]: "Alternative title" implies that there is some other "main title" that is shown. I simply change this to "Title for the dialog" or "Custom title for the dialog".

File: lms/static/scripts/frontend_apps/api-types.ts [hunk 10]
```diff
@@ -4,21 +4,25 @@
  */
 import type { APICallInfo } from './config';
 
-/**
- * Object representing a file or folder resource in the LMS's file storage.
- */
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/api-types.ts:9]: Can you update rather than remove this comment. Something like "Object representing a file, folder or document stored by the LMS".

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx [hunk 11]
```diff
@@ -333,9 +360,20 @@ export default function ContentSelector({
...
             title="Select PDF from Canvas"
           >
-            Canvas
+            Canvas File
           </OptionButton>
         )}
+        {canvasPagesEnabled && (
+          <OptionButton
+            data-testid="canvas-page-button"
+            details="PAGE"
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/ContentSelector.tsx:369]:
```suggestion
            details="Page"
```

PDF is capitalized because it is an acronym. Other non-acronym terms in this UI position are not ("Web page", "Article").

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx [hunk 12]
```diff
@@ -75,6 +75,11 @@ type LMSFilePickerProps = {
...
+   */
+  title?: string;
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx:82]: There are a few other UI strings that need to be adapted. For English, it would be adequate to provide singular and plural forms of the resource name (`file`, `files`, `page`, `pages`) and substitute those into the messages. In future we might have to provide alternative versions of the whole message to be grammatically correct.

File: lms/static/scripts/frontend_apps/components/test/ResourceList-test.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+      assert.equal(dataRow.find('td').at(1).text(), formattedDate);
+    });
+  });
+
+  it('renders a explanatory message when there are no resources', () => {
+    const wrapper = renderResourcesListNoResources({
+      isLoading: false,
+      noResourcesMessage: <div className="FileList__no-files-message" />,
+    });
+    assert.isTrue(wrapper.exists('.FileList__no-files-message'));
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/test/ResourceList-test.js:63]: This could be modernized to use `data-testid` instead.

**Review by robertknight — COMMENTED** [review]
The revised approach looks viable to me, but there are some additional UI strings that need to be updated depending on the resource type. I also had a query about sorting of files vs pages. I thought we did sort files by name, but wasn't sure where this was handled.

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx [hunk 14]
```diff
@@ -216,6 +229,20 @@ export default function ContentSelector({
...
       break;
+    case 'canvasPage':
+      // We can temporarily use the LMSFilePicker for pages
+      dialog = (
+        <LMSFilePicker
+          authToken={authToken}
+          listFilesApi={listPagesApi}
+          onCancel={cancelDialog}
+          onSelectFile={selectCanvasPage}
+          missingFilesHelpLink="TODO"
```
[the remark this answers is not in the document]
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/ContentSelector.tsx:240]: I thought Marcos mentioned something he created one, but I probably got confused with something unrelated.

I have just created a draft issue to make sure we add this: https://github.com/orgs/hypothesis/projects/126/views/1?pane=issue&itemId=40686398

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx (hunk 12 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx:82]: See my comment here https://github.com/hypothesis/lms/pull/5717#discussion_r1347090485

**Comment by acelaya** [comment]
> Pages are currently not sorted by name, which will make find an entry difficult if there are a large number. Is this handled on the frontend or backend for files?

Sorting is done server-side for files. I discussed it with Marcos and have created a [draft issue](https://github.com/orgs/hypothesis/projects/126/views/1?pane=issue&itemId=40687144) to sort pages as well.

**Comment by acelaya** [comment]
@robertknight I have addressed all non-controversial comments and added [a suggestion](https://github.com/hypothesis/lms/pull/5717#discussion_r1347090485) that would address the two unresolved comments left. Let me know what you think.

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx (hunk 8 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/ContentSelector.tsx:171]: > I'm thinking making it generic (like ResourcesList) probably makes sense, and would allow to conditionally define all the user-facing dynamic texts, without having to define a prop for each one of them.

How will the conditional definitions work? Maybe the easiest way to show what you mean will be to sketch some code and show me.

File: lms/static/scripts/frontend_apps/components/ResourceList.tsx [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+import type { ComponentChildren } from 'preact';
+
+import type { Resource } from '../api-types';
+
+export type ResourceListProps<ResourceType extends Resource> = {
+  /** List of resource objects returned by the API */
+  resources: ResourceType[];
+  /** Whether to show a loading indicator */
+  isLoading?: boolean;
+  /** The resource within `resource` which is currently selected */
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/ResourceList.tsx:17]:
```suggestion
  /** The resource within `resources` which is currently selected */
```

File: lms/static/scripts/frontend_apps/components/ResourceList.tsx [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+   */
+  onUseResource: (r: ResourceType | null) => void;
+  /** Optional message to display if there are no resources */
+  noResourcesMessage?: ComponentChildren;
+  /** Component title for accessibility. Defaults to 'File list' */
+  title?: string;
+};
+
+/**
+ * List of the resources
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/ResourceList.tsx:33]:
```suggestion
 * List of resources (files, documents etc.)
```

File: lms/static/scripts/frontend_apps/api-types.ts [hunk 17]
```diff
@@ -5,20 +5,27 @@
 import type { APICallInfo } from './config';
 
 /**
- * Object representing a file or folder resource in the LMS's file storage.
+ * Object representing a file, folder or document stored by the LMS
  */
-export type File = {
-  // FIXME - This ought to be present on all file objects.
-  type?: 'File' | 'Folder';
+export type Resource = {
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/api-types.ts:10]: Did the term "Resource" come from anywhere in particular? It's a reasonable choice, although I wonder if "Document" might be a slightly more obvious generalization of files / pages etc.

File: lms/static/scripts/frontend_apps/components/ContentSelector.tsx (hunk 8 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/ContentSelector.tsx:171]: Let me try to sketch something and push it here.

File: lms/static/scripts/frontend_apps/api-types.ts (hunk 17 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/api-types.ts:10]: Nop, nowhere specific. I just tried to think on a term that would cover Files and Pages.

I'm fine with using `Document` instead.

Comment (acelaya) [reply lms/static/scripts/frontend_apps/api-types.ts]: I have replaced any reference to `resource` with `document`, including the API type, the List component (which is now called `DocumentList`, not `ResourceList`) and any prop or comment.

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx [hunk 18]
```diff
@@ -398,16 +416,16 @@ export default function LMSFilePicker({
             <span>Unable to authorize file access.</span>
           ) : (
             <span>
-              To select a file, you must authorize Hypothesis to access your
-              files.
+              To select a {documentType}, you must authorize Hypothesis to
+              access your {documentType}s.
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx:420]: I have decided to be pragmatic, and simply use the `documentType` value (page or file) and pluralized it by appending an `s`, which is correct for those two possible values.

If this evolves into supporting more types or requiring completely different messages for every type, we will need some helper functions and adjust this.

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx (hunk 12 above)
Comment (acelaya) [reply lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx]: As [discussed on slack](https://hypothes-is.slack.com/archives/C1M8NH76X/p1696593792109319), I have refactored the `LMSFilePicker` to expose a prop that tells the typo of documents it handles.

Using that prop, all user-facing messages can be dynamically composed.

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx (hunk 18 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx:420]: We'd also need to overhaul this to support translation.

**Review by robertknight — APPROVED** [review]
LGTM
